### PR TITLE
testsys: `aws-k8s` migration testing with `cargo make test`

### PR DESCRIPTION
**Pull request by ecpullen**
**Issue number:**
#2149 

**Description of changes:**
Introduces migrations to `cargo make test`. The added documentation in `TESTING.md` will help any user set up the necessary infrastructure to run migration testing with all aws-k8s variants.

**Testing done:**
Used `cargo make -e TESTSYS_TEST=migration test` to run a migration test on aws-k8s-1.21 and all tests passed successfully.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: Makefile.toml
Comment (jpculp) [inline Makefile.toml]:
```suggestion
#    5: a final quick test on the downgraded instances
```

File: Makefile.toml
Comment (jpculp) [inline Makefile.toml]: nit. I would refer to quick test as `quick` test, just so its obvious that `quick` is the type of the quick test you are running.

File: TESTING.md
Comment (jpculp) [inline TESTING.md]: I think typically we let the markdown renderer handle the line-wrapping.

File: TESTING.md
Comment (jpculp) [inline TESTING.md]: nit. `TUF` repos

File: TESTING.md
Comment (jpculp) [inline TESTING.md]: nit. `Infra.toml`

File: tools/Cargo.lock
Comment (jpculp) [inline tools/Cargo.lock]: AWS SDK for Rust is up to `0.15.0` now _(`0.45.0` for smithy)_.

File: tools/testsys/src/aws_resources.rs
Comment (arnaldo2792) [inline tools/testsys/src/aws_resources.rs:322]: A little late to comment on this, but I just realized that in all the Bottlerocket repo, we don't use `anyhow` and instead we use `snafu`. I think we should be consistent and move at least this part of `testsys` to `snafu`. I'm not blocking on this though but an issue should be good.

File: Makefile.toml
Comment (etungsten) [inline Makefile.toml]: nit: typo
```suggestion
#    5: a final quick test on the downgraded instances
```

File: TESTING.md
Comment (etungsten) [inline TESTING.md]: nit:
```suggestion
* The AMI ID of the previous release
```

File: TESTING.md
Comment (etungsten) [reply TESTING.md]: nit: slightly different wording to hopefully make it more clear
```suggestion
Migration testing launches instances of a starting Bottlerocket version, or a provided initial AMI and migrates instances to the target version.
```

File: TESTING.md
Comment (etungsten) [inline TESTING.md:176]: Some clarification on what "migration" means in this context would be nice, i.e. to update instances and downgrade them.

File: TESTING.md
Comment (etungsten) [inline TESTING.md]: nit: more clarification
```suggestion
* Image artifacts and local TUF repos of said artifacts for current changes
```

File: TESTING.md
Comment (etungsten) [inline TESTING.md]: nit:
```suggestion
A publicly available S3 bucket is needed to complete the next steps.
```

File: TESTING.md
Comment (etungsten) [reply TESTING.md]: nit: capitalize TUF and S3

File: tools/testsys/src/aws_resources.rs
Comment (arnaldo2792) [inline tools/testsys/src/aws_resources.rs:70]: Nit: you could add a comment explaining the order of the crds

File: tools/testsys/src/aws_resources.rs
Comment (arnaldo2792) [inline tools/testsys/src/aws_resources.rs:150]: Why is the suffix no longer required?

File: tools/testsys/src/aws_resources.rs
Comment (arnaldo2792) [inline tools/testsys/src/aws_resources.rs]: Nit: comments related to each enum should be on top of the enum.

File: tools/testsys/src/aws_resources.rs
Comment (arnaldo2792) [inline tools/testsys/src/aws_resources.rs]: `starting_version` and `migrate_to_version` are both `String`, why the need of `to_string()`?

File: tools/testsys/src/aws_resources.rs
Comment (arnaldo2792) [inline tools/testsys/src/aws_resources.rs:398]: Why is the default value used here if a few lines below it will be changed to something know? The value could set here right away, correct?

File: tools/testsys/src/aws_resources.rs
Comment (arnaldo2792) [inline tools/testsys/src/aws_resources.rs]: Should this be a constant?

File: Makefile.toml
Comment (arnaldo2792) [inline Makefile.toml:119]: Please follow the convention suggested in a previous PR about the commit messages format

File: TESTING.md
Comment (arnaldo2792) [inline TESTING.md:169]: I think the two commits can be squashed in one, since they are both related to the same change.

**Comment by ecpullen** [comment]
^ addressed @jpculp comments

**Review by arnaldo2792 — CHANGES_REQUESTED** [review]

File: tools/testsys/src/aws_resources.rs
Comment (ecpullen) [reply tools/testsys/src/aws_resources.rs:398]: It is problem with rust/our templating structure. The instance_ids are a Vec in the rust struct, but we want the controller to automatically pull the instance ids from the instance provider which we do after converting from `MigrationConfig` to `Map`.

File: tools/testsys/src/aws_resources.rs
Comment (ecpullen) [reply tools/testsys/src/aws_resources.rs:322]: I will add an issue for this.

File: tools/testsys/src/aws_resources.rs
Comment (ecpullen) [reply tools/testsys/src/aws_resources.rs:150]: It was not needed anymore migration testing. It was left in case an issue in testsys (running upgrade downgrade and conformance on the same cluster) was not solved, but a solution for the issue is currently in review.

Comment (ecpullen) [reply tools/testsys/src/aws_resources.rs:150]: I can make a separate pr/commit to remove that if needed.

File: tools/testsys/src/aws_resources.rs
Comment (ecpullen) [reply tools/testsys/src/aws_resources.rs]: `format!` requires a string literal, so we can't use a constant.

**Comment by ecpullen** [comment]
^ addresses @arnaldo2792 comments

File: TESTING.md
Comment (arnaldo2792) [inline TESTING.md]: Nit: `sed 's/ami/AMI'`

File: TESTING.md
Comment (arnaldo2792) [inline TESTING.md]: Nit: `sed 's/bottlerocket/Bottlerocket'`

File: TESTING.md
Comment (arnaldo2792) [inline TESTING.md]: Nit: a link to the documentation that explains how to build and publish repos would be nice

File: TESTING.md
Comment (arnaldo2792) [inline TESTING.md]: This is not true, a contributor can use whatever they want to store their TUF repositories, so it will be best to say "the URL of a TUF repository" or something along those lines.

File: TESTING.md
Comment (arnaldo2792) [inline TESTING.md]: Nit: a line per sentence

File: TESTING.md
Comment (arnaldo2792) [inline TESTING.md]: I think there is something missing in this phrase?

```suggestion
The examples below also assume that the default repo is used in `Infra.toml`, but any repo can be used by setting the `PUBLISH_REPO` environment variable.
```

File: tools/testsys/src/aws_resources.rs
Comment (arnaldo2792) [reply tools/testsys/src/aws_resources.rs:150]: I'm fine with the update, no need for a new PR (folks may have a different point of view), but it was strange to me to update a method that was just added. I'll let you mark this conversation as resolved so that you don't miss this message.

File: tools/testsys/src/aws_resources.rs
Comment (arnaldo2792) [inline tools/testsys/src/aws_resources.rs]: There are two `println!`, are they still required?

File: TESTING.md
Comment (arnaldo2792) [inline TESTING.md]:
```suggestion
The following bash script will checkout the proper branch from git and create the build images and TUF repos for testing.
```

File: TESTING.md
Comment (arnaldo2792) [inline TESTING.md]:
```suggestion
git checkout "v${TESTSYS_STARTING_VERSION}"
```

File: TESTING.md
Comment (arnaldo2792) [inline TESTING.md]: I think all mentions of a S3 bucket should be removed, folks in the community may have their own way to update their TUF repos.

File: TESTING.md
Comment (arnaldo2792) [inline TESTING.md]: Nit:

```suggestion
git checkout "${WORKING_BRANCH}"
```

File: TESTING.md
Comment (arnaldo2792) [inline TESTING.md]: For this testing it must, right?

**Review by arnaldo2792 — CHANGES_REQUESTED** [review]

File: TESTING.md
Comment (ecpullen) [reply TESTING.md]: The example steps I provide include syncing tuf repos with the pre-update metadata. I was trying to make the process of onboarding to testsys as easy as possible by providing a sample workflow from start to finish. I can add to the beginning that it is not required to have, but this example will use it.

File: TESTING.md
Comment (ecpullen) [reply TESTING.md]: See above comment. Should we remove the step by step guide so that we can remove the s3 bucket bits?

File: TESTING.md
Comment (ecpullen) [reply TESTING.md]: For using develop yes. But if someone wanted to try building 1.6.0 and 1.6.1 to see how testsys migrations work they wouldn't need to.

File: TESTING.md
Comment (arnaldo2792) [reply TESTING.md]: I think it will be best to point whomever reads this guide to [this document](https://github.com/bottlerocket-os/bottlerocket/blob/develop/PUBLISHING.md#repo-location) so that they are aware that they are free to decide where to store their TUF repo. I would definitely remove the "publicly" available part since that isn't a good practice, and we shouldn't advocate for that. It will be best to say "use the URL of an accessible TUF repo"; and for your example, assume the contributor followed the link you will add, and that they have a URL that is on top of a properly configured S3 bucket (that's what ultimately happened in the document I linked).

File: tools/testsys/Cargo.toml
Comment (etungsten) [inline tools/testsys/Cargo.toml]: Remember to bump these to the latest version before merge.

File: TESTING.md
Comment (etungsten) [inline TESTING.md]: nit: 
```suggestion
* The AMI ID for the previous release
```

File: TESTING.md
Comment (etungsten) [inline TESTING.md]:
```suggestion
* A publicly accessible TUF repository
```

**Comment by ecpullen** [comment]
^^ Address @etungsten comments
^ rebase to develop

**Review by etungsten — APPROVED** [review]

File: Makefile.toml
Comment (bcressey) [inline Makefile.toml]: It might be useful to default to the previous version in the repository, if the starting version is not provided.

File: TESTING.md
Comment (bcressey) [inline TESTING.md]: I spotted this typo, might it be good to run the guide through a spell checker:

```suggestion
We need the URL of an accessible TUF repo so the Bottlerocket instances know where to retrieve the update metadata and targets. 
```

File: TESTING.md
Comment (bcressey) [inline TESTING.md]: This is a reasonable set of requirements, but that there's also a clear need for a tool to help meet the requirements. It's tricky enough to accomplish that ten developers will find ten different ways of doing it.

I'm not sure about taking an AMI ID. That wouldn't apply to either VMware or bare metal. I'd prefer to understand what the upgrade/downgrade solution looks like on other platforms first.

If you weren't allowed to use either AMIs or SSM parameters, how would you approach this? Can it be done with just the contents of the repo?

File: TESTING.md
Comment (bcressey) [inline TESTING.md]: We should be consistent about whether or not a `v` prefix is required, and possibly support both forms in environment variables like `TESTSYS_STARTING_VERSION`.

File: TESTING.md
Comment (bcressey) [inline TESTING.md]: This is one of the places where "`v` or not `v`" is awkward.

**Comment by ecpullen** [comment]
^ spell check on markdown and automatically determine correct AMI ID using previous bottlerocket release tag.

File: TESTING.md
Comment (bcressey) [inline TESTING.md]: Is this step still needed for something?

File: TESTING.md
Comment (bcressey) [inline TESTING.md]:
```suggestion
Remember to sync your TUF repos with the new metadata and targets.
```

File: tools/testsys/src/aws_resources.rs
Comment (bcressey) [inline tools/testsys/src/aws_resources.rs]: nit: this sort of comment seems like it could get outdated pretty quickly - it's specific in both the number and order of the tests.

It might be better to document each of the steps inline, since there's a natural separation after each `depends_on.push()` to talk about what & why each CRD is added.

File: tools/testsys/src/aws_resources.rs
Comment (bcressey) [inline tools/testsys/src/aws_resources.rs:93]: Not a big deal here but I wonder if we could do the numbering somewhere else, so it's automatically part of the name in the output based on the position in the vector.

File: tools/testsys/src/aws_resources.rs
Comment (bcressey) [inline tools/testsys/src/aws_resources.rs:99]: Also not a big deal here but should "name" be a required field or have a default value?

I wonder if the cases where the name is optional are worth the added friction in what feels like the normal case here. You could potentially have a `NamedCrd` type and add a `From` impl to `Crd` for it.

File: tools/testsys/src/aws_resources.rs
Comment (bcressey) [inline tools/testsys/src/aws_resources.rs:334]: Based on the function name I'd expect this to be providing instances. Do we need this interface or will all of the implementations basically look like this?

File: tools/testsys/src/aws_resources.rs
Comment (bcressey) [inline tools/testsys/src/aws_resources.rs:376]: This seems like a place where stronger typing could be useful. It sort of sticks out as having not very much to do with migration testing.

File: tools/testsys/src/aws_resources.rs
Comment (bcressey) [inline tools/testsys/src/aws_resources.rs]:
```suggestion
                    name: "aws-migration-test-agent".to_string(),
```

File: tools/testsys/src/run.rs
Comment (bcressey) [inline tools/testsys/src/run.rs]: nit: comment is outdated

File: tools/testsys/src/run.rs
Comment (bcressey) [inline tools/testsys/src/run.rs]: nit: I'd either remove these "bottlerocket" references or capitalize them

```suggestion
    /// The starting version for migrations. This is required for all migrations tests.
```

File: tools/testsys/src/run.rs
Comment (bcressey) [inline tools/testsys/src/run.rs]: nit: this comment can just be a high-level summary of what migration testing is, rather than the required arguments or the specific test sequence.

File: TESTING.md
Comment (bcressey) [reply TESTING.md]: The current implementation addresses my concern here, thanks!

By automating the lookup of the previous version and the AMI, it's more likely that the developer will follow the expected steps rather than going "off script" to force the use of AMIs built some other way.

File: tools/testsys/src/aws_resources.rs
Comment (ecpullen) [reply tools/testsys/src/aws_resources.rs:334]: This will be used for the ecs and VMware workflow as well. It is just the name of the resource object responsible for providing instances.

File: tools/testsys/src/aws_resources.rs
Comment (ecpullen) [reply tools/testsys/src/aws_resources.rs:376]: I was wondering if an instance provider trait would be helpful or not.

File: tools/testsys/src/aws_resources.rs
Comment (ecpullen) [reply tools/testsys/src/aws_resources.rs:99]: The crd itself (for tests and resources) has the name field as an `Option` so somewhere there needs to be an `unwrap()` or `context()` to read it unless we default with  `""`, but that has the potential to cause issues if the field was missing for some reason.

**Review by bcressey — APPROVED** [review]

**Review by arnaldo2792 — APPROVED** [review]

**Pull request merged**
